contracts/receivables/RealWorldReceivable.sol

```diff
@@ -0,0 +1,189 @@
+        if (paymentToken != address(0) && currencyCode != 0)
+            revert Errors.multipleCurrenciesGiven();
+
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+        _safeMint(msg.sender, tokenId);
+
+        RealWorldReceivableInfo memory receivableInfo = RealWorldReceivableInfo(
+            poolAddress,
+            paymentToken,
+            receivableAmount,
+            0, // paidAmount
+            maturityDate,
+            uint64(block.timestamp),
+            currencyCode
+        );
+        receivableInfoMapping[tokenId] = receivableInfo;
+
+        _setTokenURI(tokenId, uri);
+    }
+
+    /**
+     * @dev Declares payment for a receivable.
+     * Only the owner of the token can declare a payment.
+     * The payment method for the receivable must be Declarative.
+     * The receivable must not already be paid in full.
+     * Emits a `PaymentDeclared` event.
+     * @param tokenId The ID of the receivable token.
+     * @param paymentAmount The amount of payment being declared.
+     */
+    function declarePayment(uint256 tokenId, uint96 paymentAmount) public {
+        if (msg.sender != ownerOf(tokenId)) revert Errors.notNFTOwner();
+        RealWorldReceivableInfo storage receivableInfo = receivableInfoMapping[tokenId];
+        receivableInfo.paidAmount += paymentAmount;
+
+        emit PaymentDeclared(
+            msg.sender,
+            receivableInfo.poolAddress,
+            tokenId,
+            receivableInfo.paymentToken,
+            receivableInfo.currencyCode,
+            uint256(paymentAmount)
+        );
+    }
+
+    /**
+     * @dev Gets the payment status of a receivable.
+     * Returns `Status.Paid` if the receivable has been paid in full.
+     * Returns `Status.PartiallyPaid` if the receivable has been paid partially.
+     * Returns `Status.Unpaid` if the receivable has not been paid at all.
+     * @param tokenId The ID of the receivable token.
+     * @return The payment status of the receivable.
+     */
+    function getStatus(uint256 tokenId) public view returns (Status) {
+        RealWorldReceivableInfo memory receivableInfo = receivableInfoMapping[tokenId];
+        if (receivableInfo.paidAmount >= receivableInfo.receivableAmount) {
+            return Status.Paid;
+        } else if (receivableInfo.paidAmount > 0) {
+            return Status.PartiallyPaid;
+        } else {
+            return Status.Unpaid;
+        }
+    }
+
+    // The following functions are overrides required by Solidity.
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId,
+        uint256 batchSize
+    ) internal override(ERC721Upgradeable, ERC721EnumerableUpgradeable) {
+        super._beforeTokenTransfer(from, to, tokenId, batchSize);
+    }
+
+    function _burn(uint256 tokenId)
+        internal
+        override(ERC721Upgradeable, ERC721URIStorageUpgradeable)
+    {
+        super._burn(tokenId);
```

Comment:
Should we prevent a receivable to be burned if it is not marked as paid?

Reply:
In my mind burning the receivable would satisfy the need for a "Voided" status. What do you think?

Reply:
Sounds good not to add the additional check.